File: typecast-python/src/typecast/composer.py
from __future__ import annotations

import math
import re
from dataclasses import dataclass
from typing import Callable, Literal, Optional, Union

from .models import LanguageCode, Output, TTSModel, TTSPrompt, TTSRequest, TTSResponse

_PAUSE_TOKEN = re.compile(r"<\|(\d+(?:\.\d+)?)s\|>")


@dataclass(frozen=True)
class _TextPart:
    kind: Literal["text"]
    text: str


@dataclass(frozen=True)
class _PausePart:
    kind: Literal["pause"]
    seconds: float


@dataclass(frozen=True)
class _SpeechPart:
    text: str
    settings: "_ComposerSettings"


@dataclass(frozen=True)
class _ComposerSettings:
    voice_id: Optional[str] = None
    model: Optional[Union[TTSModel, str]] = None
    language: Optional[Union[LanguageCode, str]] = None
    prompt: Optional[TTSPrompt] = None
    output: Optional[Output] = None
    seed: Optional[int] = None


class SpeechComposer:
    def __init__(self, compose: Callable[[list[dict]], TTSResponse]):
        self._compose = compose
        self._defaults = _ComposerSettings()
        self._parts: list[Union[_SpeechPart, _PausePart]] = []

    def defaults(
        self,
        *,
        voice_id: Optional[str] = None,
        model: Optional[Union[TTSModel, str]] = None,
        language: Optional[Union[LanguageCode, str]] = None,
        prompt: Optional[TTSPrompt] = None,
        output: Optional[Output] = None,
        seed: Optional[int] = None,
    ) -> "SpeechComposer":
        self._defaults = _merge_settings(
            self._defaults,
            _ComposerSettings(
                voice_id=voice_id,
                model=model,
                language=language,
                prompt=prompt,
                output=output,
                seed=seed,
            ),
        )
        return self

    def say(
        self,
        text: str,
        *,
        voice_id: Optional[str] = None,
        model: Optional[Union[TTSModel, str]] = None,
        language: Optional[Union[LanguageCode, str]] = None,
        prompt: Optional[TTSPrompt] = None,
        output: Optional[Output] = None,
        seed: Optional[int] = None,
    ) -> "SpeechComposer":
        settings = _merge_settings(
            self._defaults,
            _ComposerSettings(
                voice_id=voice_id,
                model=model,
                language=language,
                prompt=prompt,
                output=output,
                seed=seed,
            ),
        )
        self._parts.append(_SpeechPart(text=text, settings=settings))
        return self

    def pause(self, seconds: float) -> "SpeechComposer":
        """Insert silence between speech segments.

        Args:
            seconds: Duration in seconds. Use 0.3 for 300 ms, 3 for 3 seconds.
        """
        if not math.isfinite(seconds) or seconds <= 0:
            raise ValueError("pause seconds must be greater than 0")
        self._parts.append(_PausePart(kind="pause", seconds=seconds))
        return self

    def generate(self) -> TTSResponse:
        plan = self._build_plan()
        if not any(isinstance(part, _SpeechPart) for part in plan):
            raise ValueError("at least one speech segment is required")

        output_format = (
            self._defaults.output.audio_format if self._defaults.output else "wav"
        )
        if output_format not in ("wav", "mp3"):
            raise ValueError(
                f"unsupported composed speech output format: {output_format}"
            )

        segments: list[dict] = []
        for part in plan:
            if isinstance(part, _PausePart):
                segments.append({"type": "pause", "duration_seconds": part.seconds})
                continue
            request = _settings_to_request(part.text, part.settings, output_format)
            segments.append({"type": "tts", **request.model_dump(exclude_none=True)})
        return self._compose(segments)

    def _build_plan(self) -> list[Union[_SpeechPart, _PausePart]]:
        plan: list[Union[_SpeechPart, _PausePart]] = []
        for part in self._parts:
            if isinstance(part, _PausePart):
                plan.append(part)
                continue
            for parsed in parse_pause_markup(part.text):
                if isinstance(parsed, _PausePart):
                    plan.append(parsed)
                    continue
                if not parsed.text.strip():
                    continue
                if not part.settings.voice_id:
                    raise ValueError(
                        "voice_id is required for composed speech segments"
                    )
                if not part.settings.model:
                    raise ValueError("model is required for composed speech segments")
                plan.append(_SpeechPart(text=parsed.text, settings=part.settings))
        return plan


def parse_pause_markup(text: str) -> list[Union[_TextPart, _PausePart]]:
    parts: list[Union[_TextPart, _PausePart]] = []
    last_index = 0
    for match in _PAUSE_TOKEN.finditer(text):
        if match.start() > last_index:
            parts.append(_TextPart(kind="text", text=text[last_index : match.start()]))
        parts.append(_PausePart(kind="pause", seconds=float(match.group(1))))
        last_index = match.end()
    if last_index < len(text):
        parts.append(_TextPart(kind="text", text=text[last_index:]))
    return parts


def _merge_settings(
    base: _ComposerSettings,
    override: _ComposerSettings,
) -> _ComposerSettings:
    return _ComposerSettings(
        voice_id=override.voice_id if override.voice_id is not None else base.voice_id,
        model=override.model if override.model is not None else base.model,
        language=override.language if override.language is not None else base.language,
        prompt=override.prompt if override.prompt is not None else base.prompt,
        output=_merge_output(base.output, override.output),
        seed=override.seed if override.seed is not None else base.seed,
    )


def _merge_output(
    base: Optional[Output], override: Optional[Output]
) -> Optional[Output]:
    if base is None and override is None:
        return None
    data = (
        base.model_dump(exclude_none=True, exclude_unset=True)
        if base is not None
        else {}
    )
    if override is not None:
        data.update(override.model_dump(exclude_none=True, exclude_unset=True))
    return Output(**data)


def _settings_to_request(
    text: str, settings: _ComposerSettings, output_format: str
) -> TTSRequest:
    output = _merge_output(settings.output, Output(audio_format=output_format))
    return TTSRequest(
        text=text,
        voice_id=settings.voice_id or "",
        model=settings.model
        if isinstance(settings.model, TTSModel)
        else TTSModel(settings.model),
        language=settings.language,
        prompt=settings.prompt,
        output=output,
        seed=settings.seed,
    )

File: typecast-js/src/composer.ts
import type { Output, TTSRequest, TTSResponse } from './types/TextToSpeech';

export type ComposeSegment =
  ({ type: 'tts' } & TTSRequest) | { type: 'pause'; duration_seconds: number };
type ComposeFn = (segments: ComposeSegment[]) => Promise<TTSResponse>;
type ComposerSettings = Omit<Partial<TTSRequest>, 'text'>;

type SpeechPart =
  | { kind: 'speech'; text: string; overrides?: ComposerSettings }
  | { kind: 'pause'; seconds: number };

type ParsedPart = { kind: 'text'; text: string } | { kind: 'pause'; seconds: number };

const PAUSE_TOKEN = /<\|(\d+(?:\.\d+)?)s\|>/g;
export class SpeechComposer {
  private defaultSettings: ComposerSettings = {};
  private parts: SpeechPart[] = [];

  constructor(private readonly compose: ComposeFn) {}

  defaults(settings: ComposerSettings): this {
    this.defaultSettings = mergeSettings(this.defaultSettings, settings);
    return this;
  }

  say(text: string, overrides?: ComposerSettings): this {
    this.parts.push({ kind: 'speech', text, overrides });
    return this;
  }

  /**
   * Inserts silence between speech segments.
   *
   * @param seconds Duration in seconds. Examples: 0.3 for 300 ms, 3 for 3 seconds.
   */
  pause(seconds: number): this {
    if (!Number.isFinite(seconds) || seconds <= 0) {
      throw new Error('pause seconds must be greater than 0');
    }
    this.parts.push({ kind: 'pause', seconds });
    return this;
  }

  async generate(): Promise<TTSResponse> {
    const outputFormat = this.resolveOutputFormat();
    const plan = this.buildPlan(outputFormat);
    if (!plan.some((part) => part.kind === 'speech')) {
      throw new Error('at least one speech segment is required');
    }

    return this.compose(
      plan.map((part) =>
        part.kind === 'pause'
          ? { type: 'pause', duration_seconds: part.seconds }
          : { type: 'tts', ...part.request },
      ),
    );
  }

  private resolveOutputFormat(): 'wav' | 'mp3' {
    const formats = new Set<'wav' | 'mp3'>();
    for (const part of this.parts) {
      if (part.kind !== 'speech' || !part.text.trim()) continue;
      const format = mergeSettings(this.defaultSettings, part.overrides ?? {}).output?.audio_format;
      if (format) formats.add(format);
    }
    if (formats.size > 1) throw new Error('composed speech segments must use one audio format');
    return formats.values().next().value ?? 'wav';
  }

  private buildPlan(
    outputFormat: 'wav' | 'mp3',
  ): Array<{ kind: 'speech'; request: TTSRequest } | { kind: 'pause'; seconds: number }> {
    const plan: Array<
      { kind: 'speech'; request: TTSRequest } | { kind: 'pause'; seconds: number }
    > = [];
    for (const part of this.parts) {
      if (part.kind === 'pause') {
        plan.push(part);
        continue;
      }

      const settings = mergeSettings(this.defaultSettings, part.overrides ?? {});
      for (const parsed of parsePauseMarkup(part.text)) {
        if (parsed.kind === 'pause') {
          plan.push(parsed);
          continue;
        }
        if (!parsed.text.trim()) {
          continue;
        }
        if (!settings.voice_id) {
          throw new Error('voice_id is required for composed speech segments');
        }
        if (!settings.model) {
          throw new Error('model is required for composed speech segments');
        }
        plan.push({
          kind: 'speech',
          request: {
            ...settings,
            text: parsed.text,
            voice_id: settings.voice_id,
            model: settings.model,
            output: {
              ...settings.output,
              audio_format: outputFormat,
            },
          },
        });
      }
    }
    return plan;
  }
}

export function parsePauseMarkup(text: string): ParsedPart[] {
  const parts: ParsedPart[] = [];
  let lastIndex = 0;
  for (const match of text.matchAll(PAUSE_TOKEN)) {
    const index = match.index;
    const seconds = Number(match[1]);
    if (!Number.isFinite(seconds) || seconds <= 0) continue;
    if (index > lastIndex) {
      parts.push({ kind: 'text', text: text.slice(lastIndex, index) });
    }
    parts.push({ kind: 'pause', seconds });
    lastIndex = index + match[0].length;
  }
  if (lastIndex < text.length) {
    parts.push({ kind: 'text', text: text.slice(lastIndex) });
  }
  return parts;
}

function mergeSettings(base: ComposerSettings, override: ComposerSettings): ComposerSettings {
  return {
    ...base,
    ...override,
    output: mergeOutput(base.output, override.output),
  };
}

function mergeOutput(base?: Output, override?: Output): Output | undefined {
  if (!base && !override) return undefined;
  return { ...(base ?? {}), ...(override ?? {}) };
}
